mlflow/sagemaker/__init__.py

```diff
@@ -354,14 +511,148 @@ def _deploy(role, image_url, app_name, model_s3_path, run_id, region_name):
         Tags=[
             {
                 'Key': 'app_name',
-                'Value': app_name,
+                'Value': endpoint_name,
             },
         ],
     )
-    eprint("endpoint_config_arn: %s" % endpoint_config_response["EndpointConfigArn"])
+    
+    eprint("Created endpoint configuration with arn: %s" 
+            % endpoint_config_response["EndpointConfigArn"])
     endpoint_response = sage_client.create_endpoint(
-        EndpointName=app_name,
+        EndpointName=endpoint_name,
         EndpointConfigName=config_name,
         Tags=[],
     )
-    eprint("endpoint_arn: %s" % endpoint_response["EndpointArn"])
+    eprint("Created endpoint with arn: %s" % endpoint_response["EndpointArn"])
+
+def _update_sagemaker_endpoint(endpoint_name, 
+                               image_url, 
+                               model_s3_path, 
+                               run_id, 
+                               mode, 
+                               archive, 
+                               role, 
+                               sage_client, 
+                               s3_client):
+    """
+    :param image_url: URL of the ECR-hosted docker image the model is being deployed into
+    :param model_s3_path: s3 path where we stored the model artifacts
+    :param run_id: RunId that generated this model
+    :param mode: either mlflow.sagemaker.DEPLOYMENT_MODE_ADD or 
+                 mlflow.sagemaker.DEPLOYMENT_MODE_REPLACE.
+    :param archive: If True, any pre-existing SageMaker application resources that become inactive
+                    (i.e. as a result of deploying in mlflow.sagemaker.DEPLOYMENT_MODE_REPLACE mode) 
+                    will be preserved. If False, these resources will be deleted.
+    :param role: SageMaker execution ARN role
+    :param sage_client: A boto3 client for SageMaker
+    :param s3_client: A boto3 client for S3 
+    """
+    if mode not in [DEPLOYMENT_MODE_ADD, DEPLOYMENT_MODE_REPLACE]:
+        msg = "Invalid mode `{md}` for deployment to a pre-existing application".format(md=mode)
+        raise Exception(msg)
+   
+    endpoint_info = sage_client.describe_endpoint(EndpointName=endpoint_name)
+    endpoint_arn = endpoint_info["EndpointArn"]
+    deployed_config_name = endpoint_info["EndpointConfigName"]
+    deployed_config_info = sage_client.describe_endpoint_config(
+            EndpointConfigName=deployed_config_name)
+    deployed_config_arn = deployed_config_info["EndpointConfigArn"]
+    deployed_production_variants = deployed_config_info["ProductionVariants"]
+
+    eprint("Found active endpoint with arn: {earn}. Updating...".format(earn=endpoint_arn))
+
+    new_model_name = _get_sagemaker_model_name(endpoint_name)
```

Comment:
could create a helper `_create_sagemaker_model` since the logic is repeated twice

Reply:
Done